12/umbraco-cms/reference/routing/custom-middleware.md

```diff
@@ -0,0 +1,70 @@
+---
+description: "Customizing the ASP.NET middleware pipeline in Umbraco"
+---
+
+# Middleware
+
+Middleware is a way of handling requests and responses before or after they are completed. An example can be checking if a user is logged in before adding an item to the cart. In this example it checks something before an http request is made.
+
+In Umbraco we configure some middleware by default and in a specific order based on the [Microsoft Documentation](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-7.0#middleware-order).
+
+You are also able to add your own in-between middleware by using the Umbraco pipeline filters. In addtion you can also add settings such as: the `PrePipeline`, `PostPipeline`, `PreRouting`, `PostRouting`, and [`Endpoints`](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-7.0#endpoints) callbacks.
+
+- `PrePipeline` occurs as an early stage of request pipeline and an influence how requests are processed. An example of using Prepipeline can be for [URL rewrites](iisrewriterules.md)
+- `PostPipeline`renders content. By using middleware, the response of content can be changed before being rendered.
+- `PreRouting`can be used to change the incoming URL/definers which URL should be served for the content
+- `PostRouting`can be used to change how the content should be displayed based on routing parameters.
+- `Endpoints` is the final step of the request pipeline where the content is rendered and returned on the requested browser.
+
+The addition of the `PreRouting` and `PostRouting` is to allow correctly configuring the Cross-Origin Resource Sharing (CORS) middleware. This is done using the `IUmbracoPipelineFilters` without having to use the `WithCustomMiddleware()`.
+
+- `IUmbracoPipelineFilter` is an interface in Umbraco that allows the creation of custom filters which then modifies the behavior of the request pipeline. It can be used to change different aspects of how Umbraco handles incoming requests, such as changing content or adding security checks.
+- `WithCustomMiddleware()` is a method that can be used in Umbraco for adding custom middleware. This includes some specific customizable instructions that run in the request processing pipeline. This method is often used in combination with IUmbracoPipelineFilter.
+
+{% hint style="warning" %}
+
+`WithCustomMiddleware` should only be used as a last resort, as Umbraco can break if you forget to add middleware or add them in the wrong order.  
```

Comment:
```suggestion
Using `WithCustomMiddleware()` instead of `WithMiddelware()` should only be used as a last resort, as Umbraco can break if you forget to add middleware or add them in the wrong order.
```